src/bin/server/stream.rs

```diff
@@ -0,0 +1,227 @@
+struct ChunkedStreamReader<R>
+where
+    R: Read + Seek + Send + 'static,
+{
+    /// Container for the stream. This is used to manually keep track of the
+    /// lifetime of the reader.
+    stream_option: Option<R>,
+
+    /// Future which resolves once a chunk is read.
+    future: Option<CpuFuture<(R, Bytes), io::Error>>,
+
+    /// Pool which reading from the reader will be executed on.
+    cpu_pool: CpuPool,
+
+    /// When the offset is greater than or equal to this value, stop reading from
+    /// the stream.
+    upper_bound: u64,
+
+    /// Number of bytes from the beginning of the file the next read should start
+    /// from.
+    offset: u64,
+
+    /// Maximum size of each chunk. The last chunk may be shorter.
+    chunk_size: u64,
+}
+
+impl<R> ChunkedStreamReader<R>
+where
+    R: Read + Seek + Send + 'static,
+{
+    fn new(stream: R, range: HttpRange, cpu_pool: CpuPool) -> ChunkedStreamReader<R> {
+        ChunkedStreamReader {
+            stream_option: Some(stream),
+            future: None,
+            cpu_pool,
+            upper_bound: range.start + range.length,
+            offset: range.start,
+            chunk_size: u16::max_value() as u64,
+        }
+    }
+
+    /// Waits for the current future chunk of bytes to resolve, then returns it. Panics if there is
+    /// no future set.
+    fn wait_for_future(&mut self) -> Poll<Option<Bytes>, io::Error> {
+        let (stream, bytes) = try_ready!(self.future.as_mut().expect("no future set").poll());
+
+        self.future.take();
+        self.stream_option = Some(stream);
+        self.offset += bytes.len() as u64;
+
+        Ok(Async::Ready(Some(bytes)))
+    }
+}
+
+impl<R> Stream for ChunkedStreamReader<R>
+where
+    R: Read + Seek + Send + 'static,
+{
+    type Item = Bytes;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<Bytes>, io::Error> {
+        if self.future.is_some() {
+            return self.wait_for_future();
+        }
+
+        if self.offset >= self.upper_bound {
+            // End of stream.
+            return Ok(Async::Ready(None));
+        }
+
+        // Pick values off self to move into the closure to execute on the
+        // cpu_pool.
+        let upper_bound = self.upper_bound;
+        let offset = self.offset;
+        let chunk_size = self.chunk_size;
+        let mut inner_stream = self
+            .stream_option
+            .take()
+            .expect("stream used after end of stream");
+
+        self.future = Some(self.cpu_pool.spawn_fn(move || {
+            let chunk_size = u64::min(upper_bound - offset, chunk_size) as usize;
+            let mut buffer = BytesMut::with_capacity(chunk_size);
+
+            inner_stream.seek(SeekFrom::Start(offset))?;
+            let bytes_read = inner_stream.read(unsafe { buffer.bytes_mut() })?;
+            if bytes_read == 0 {
+                return Err(io::ErrorKind::UnexpectedEof.into());
+            }
+            unsafe { buffer.advance_mut(bytes_read) };
```

Comment:
These unsafe blocks are still not needed. Replace them with safe versions. See previous comments.

Reply:
The first use of unsafe can be made safe by using `Vec<u8>` instead of `BytesMut`, because we gain zero benefit from using `BytesMut`. When the bytes version is initalized, it just creates a `Vec<u8>` internally. Initializing a zero-filled vector is very cheap ([it uses `RawVec::with_capacity_zeroed`, which is basically just caling `calloc`](https://github.com/rust-lang/rust/blob/54d82d0880f93e293bc8fd6439ccfa7970d41f02/src/liballoc/vec.rs#L1597-L1613)).
```rust
let mut buffer = vec![0; chunk_size];
```
Now that we are using a vector, the first unsafe line can be replaced with:
```rust
let bytes_read = inner_stream.read(&mut buffer)?;
```

The second unsafe line is not even needed. What it does is advance the internal cursor used for writing to the buffer. Since the buffer is never written to after that line, it is unnecessary and the line can be removed.

Now there are no unsafe lines.